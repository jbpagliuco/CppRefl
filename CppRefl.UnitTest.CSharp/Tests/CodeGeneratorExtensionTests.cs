using CppRefl.CodeGeneration;
using CppRefl.CodeGeneration.Reflection;

namespace CppRefl.UnitTest.CSharp.Tests
{
	public class TestCodeGenerator : ICodeGeneratorExtension
	{
		public void WriteHeaderTop(CodeWriter writer, Registry registry)
		{
			writer.WriteLine("ClassFileStart");
		}

		public void WriteHeaderBottom(CodeWriter writer, Registry registry)
		{
			writer.WriteLine("ClassFileEnd");
		}

		public void WriteClassDeclaration(CodeWriter writer, ClassInfo classInfo, Registry registry)
		{
			writer.WriteLine($"Class: {classInfo.Type.Name}");
		}
	}

	internal class CodeGeneratorExtensionTests
	{
		//private Compiler.Compiler Compiler { get; } = new(TestUtil.CompilerParams);
		//private Registry Registry => Compiler.Registry;

		//private CodeGenerator CodeGenerator { get; set; } = null!;

		//private string OutputDirectory => Path.Combine(Path.GetTempPath(), nameof(CodeGeneratorExtensionTests));

		//[OneTimeSetUp]
		//public void Setup()
		//{
		//	Compiler.GenerateRegistry();

		//	CodeGenerator.LoadCodeGenerator(TestUtil.Filepaths.UnitTestCSharpDll);

		//	if (Directory.Exists(OutputDirectory))
		//	{
		//		Directory.Delete(OutputDirectory, true);
		//	}

		//	CodeGenerator = new(new CodeGeneratorParams()
		//	{
		//		Registry = Registry,
		//		InputFilename = TestUtil.Filepaths.ReflectedCodeFile,
		//		ModuleDirectory = TestUtil.Filepaths.UnitTestCppDirectory,
		//		OutputDirectory = OutputDirectory
		//	});
		//	CodeGenerator.GenerateCode();
		//}

		//[Test]
		//public void Test()
		//{
		//	string fileContents = File.ReadAllText(Path.Combine(OutputDirectory, $"ReflectedCode.reflgen.h")).Replace("\r\n", "\n");
		//	string expectedContents = File.ReadAllText(Path.Combine(TestUtil.Filepaths.UnitTestCSharpDirectory, "Tests", $"{nameof(CodeGeneratorExtensionTests)}.Expected.txt")).Replace("\r\n", "\n");
		//	Assert.That(fileContents, Is.EqualTo(expectedContents));
		//}
	}
}
