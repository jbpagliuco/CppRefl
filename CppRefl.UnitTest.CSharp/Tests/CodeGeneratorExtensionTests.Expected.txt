// This must be defined before we include the private macros so GENERATED_REFLECTION_CODE() works correctly.
#undef CPPREFL_INTERNAL_FILE_ID
#define CPPREFL_INTERNAL_FILE_ID ReflectedCode


// Header guard. Raise an error if this file is included multiple times.
#if defined(ReflectedCode_REFLGEN_H)
#error Including ReflectedCode.reflgen.h multiple times! Use `#pragma once` in ReflectedCode.h.
#endif
#define ReflectedCode_REFLGEN_H


#include "Private/CppReflGeneratedCodeMacros.h"



ClassFileStart


////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////
// PrimitiveTypes

#if !defined(CPPREFL_BUILD_REFLECTION)
// Macro to be added inside the definition of a reflected class.
#define GENERATED_REFLECTION_CODE_ReflectedCode9() \

	public:\
		static constexpr const char* CLASS_QUALIFIED_NAME	= "PrimitiveTypes";\
		static constexpr const char* CLASS_NAME				= CLASS_QUALIFIED_NAME + 0;\
		\
	Class: PrimitiveTypes\
		\

#else // !defined(CPPREFL_BUILD_REFLECTION)

// Define empty macro when building reflection.
#define GENERATED_REFLECTION_CODE_ReflectedCode9()

#endif // !defined(CPPREFL_BUILD_REFLECTION)

////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////


////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////
// ReflectedClass

#if !defined(CPPREFL_BUILD_REFLECTION)
// Macro to be added inside the definition of a reflected class.
#define GENERATED_REFLECTION_CODE_ReflectedCode30() \

	public:\
		static constexpr const char* CLASS_QUALIFIED_NAME	= "ReflectedClass";\
		static constexpr const char* CLASS_NAME				= CLASS_QUALIFIED_NAME + 0;\
		\
	Class: ReflectedClass\
		\

#else // !defined(CPPREFL_BUILD_REFLECTION)

// Define empty macro when building reflection.
#define GENERATED_REFLECTION_CODE_ReflectedCode30()

#endif // !defined(CPPREFL_BUILD_REFLECTION)

////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////


////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////
// ClassWithAttributes

#if !defined(CPPREFL_BUILD_REFLECTION)
// Macro to be added inside the definition of a reflected class.
#define GENERATED_REFLECTION_CODE_ReflectedCode53() \

	public:\
		static constexpr const char* CLASS_QUALIFIED_NAME	= "ClassWithAttributes";\
		static constexpr const char* CLASS_NAME				= CLASS_QUALIFIED_NAME + 0;\
		\
	Class: ClassWithAttributes\
		\

#else // !defined(CPPREFL_BUILD_REFLECTION)

// Define empty macro when building reflection.
#define GENERATED_REFLECTION_CODE_ReflectedCode53()

#endif // !defined(CPPREFL_BUILD_REFLECTION)

////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////


////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////
// ReflectedStruct

#if !defined(CPPREFL_BUILD_REFLECTION)
// Macro to be added inside the definition of a reflected class.
#define GENERATED_REFLECTION_CODE_ReflectedCode70() \

	public:\
		static constexpr const char* CLASS_QUALIFIED_NAME	= "ReflectedStruct";\
		static constexpr const char* CLASS_NAME				= CLASS_QUALIFIED_NAME + 0;\
		\
	Class: ReflectedStruct\
		\

#else // !defined(CPPREFL_BUILD_REFLECTION)

// Define empty macro when building reflection.
#define GENERATED_REFLECTION_CODE_ReflectedCode70()

#endif // !defined(CPPREFL_BUILD_REFLECTION)

////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////


////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////
// TestNamespace::ReflectedClass

#if !defined(CPPREFL_BUILD_REFLECTION)
// Macro to be added inside the definition of a reflected class.
#define GENERATED_REFLECTION_CODE_ReflectedCode84() \

	public:\
		static constexpr const char* CLASS_QUALIFIED_NAME	= "TestNamespace::ReflectedClass";\
		static constexpr const char* CLASS_NAME				= CLASS_QUALIFIED_NAME + 15;\
		\
	Class: TestNamespace::ReflectedClass\
		\

#else // !defined(CPPREFL_BUILD_REFLECTION)

// Define empty macro when building reflection.
#define GENERATED_REFLECTION_CODE_ReflectedCode84()

#endif // !defined(CPPREFL_BUILD_REFLECTION)

////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////

ClassFileEnd

